PWAWidgets/README.md

```diff
@@ -0,0 +1,975 @@
+   1. Else set <var>event["action"]</var> to the user action bound to <var>message</var>.
+   1. Let <var>instanceId</var> be the id of the Widget Instance bound to <var>message</var>.
+   1. Set <var>event["instance"]</var> to <var>instanceId</var>.
+   1. Let <var>widget</var> be the result of running the algorithm specified in [getByInstance(instanceId)](#widgetsgetbyinstance) with <var>instanceId</var>.
+   1. Set <var>event["tag"]</var> to <var>widget["tag"]</var>.
+   1. If <var>message</var> includes bound data,
+      1. Set <var>event["data"]</var> to the data value bound to <var>message</var>.
+1. Return <var>event</var>
+
+
+### WidgetInstall
+
+When the User Agent receives a request to create a new instance of a widget, it will need to create a placeholder for the instance before triggering the WidgetClick event within the Service Worker.
+
+Required `WidgetEvent` data:
+
+* `host`
+* `instance`
+* `tag`
+
+The <b id="creating-a-placeholder-instance">steps for creating a placeholder instance</b> with WidgetEvent <var>event</var>:
+
+1. Let <var>tag</var> be <var>event["widget"]["tag"]</var>.
+1. Let <var>widget</var> be the result of running the algorithm specified in [getByTag(tag)](#widgetsgetbytag) with <var>tag</var>.
+1. If <var>widget</var> is undefined, exit.
+1. Let <var>payload</var> be an object.
+1. Set <var>payload["data"]</var> to an empty JSON object.
+1. Set <var>payload["settings"]</var> to the result of [creating a default `WidgetSettings` object](#creating-a-default-widgetsettings-object) with <var>widget</var>.
+1. Set <var>payload["definition"]</var> to the value of <var>widget["definition"]</var>.
+1. Set <var>payload</var> to the result of [injecting manifest members into a `WidgetPayload`](#injecting-manifest-members-into-a-payload) with <var>payload</var>.
+1. Let <var>instance</var> be the result of [creating an instance](#creating-a-widget-instance) with <var>event["widget"]["id"]</var>, <var>event["widget"]["host"]</var>, and <var>payload</var>.
+1. Append <var>instance</var> to <var>widget["instances"]</var>.
+
+<p id="install">Here is the flow for install:</p>
+
+![](media/install.gif)
+
+1. A "WidgetInstall" signal is received by the User Agent, the placeholder instance is created, and the event is passed along to the Service Worker.
+2. The Service Worker
+    a. captures the Widget Instance `id` from the `widget` property,
+    b. looks up the Widget via `widgets.getByInstance()`, and
+    c. makes a `Request` for its `data` endpoint.
+3. The Service Worker then combines the `Response` with the Widget definition and passes that along to the [Widget Service](#dfn-widget-service) via the `updateInstance()` method.
+
+### WidgetUninstall
+
+Required `WidgetEvent` data:
+
+* `host`
+* `instance`
+* `tag`
+
+<p id="uninstall">The "uninstall" process is similar:</p>
+
+![](media/uninstall.gif)
```

Comment:
And here too

Reply:
The text below fully describes the image, which is why I didn’t include it.